logger/src/metrics.rs

```diff
@@ -116,6 +116,58 @@ impl Serialize for SharedMetric {
     }
 }
 
+/// Representation of a metric that is expected to be incremented from more than one thread, and
+/// which can be only 0 or 1 (i.e. is boolean), and which resets to 0 whenever it is serialised.
+#[derive(Default)]
+pub struct ResettingBooleanSharedMetric(AtomicBool);
+
+impl Metric for ResettingBooleanSharedMetric {
+    fn add(&self, value: usize) {
+        if value > 0 {
+            self.0.store(true, Ordering::Relaxed);
+        }
+    }
+
+    fn count(&self) -> usize {
+        // The Rust Standard specifies that converting a boolean to an integer result in
+        // either 0 or 1, as we want for this application.
+        self.0.load(Ordering::Relaxed) as usize
+    }
+}
+
+impl Serialize for ResettingBooleanSharedMetric {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let res = serializer.serialize_u64(self.count() as u64);
+        self.0.store(false, Ordering::Relaxed);
```

Comment:
Serialization is not atomic in this implementation. The metric could be incremented between this instruction and the one above and we will lose an event/metric.

Atomic primitives usually have mechanisms for checking and updating their value in the same operation. Take a look at [compare_and_swap](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html#method.compare_and_swap) or [compare_exchange](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html#method.compare_exchange).